gh-119127: functools.partial placeholders (#119827)
python/cpython: Doc/library/functools.rst

.. function:: partial(func, /, *args, **keywords)

>    Return a new :ref:`partial object <partial-objects>` which when called

Reviewer: Unrelated change.

   will behave like *func* called with the positional arguments *args*
   and keyword arguments *keywords*. If more arguments are supplied to the
   call, they are appended to *args*. If additional keyword arguments are
   supplied, they extend and override *keywords*.
   Roughly equivalent to::

>       Placeholder = object()

Reviewer: It is not used in the code below.

.. function:: partial(func, /, *args, **keywords)

   Return a new :ref:`partial object <partial-objects>` which when called
   will behave like *func* called with the positional arguments *args*
   and keyword arguments *keywords*. If more arguments are supplied to the
   call, they are appended to *args*. If additional keyword arguments are
   supplied, they extend and override *keywords*.
   Roughly equivalent to::

      Placeholder = object()

      def partial(func, /, *args, **keywords):

>           # Trim trailing placeholders and count how many remaining

Reviewer 2: I suggest keeping the current rough equivalent and just adding some examples that use `Partial`. The full equivalent is too messy to be useful in the main docs. Also, we already link to the source code from the top of the documentation page.

   supplied, they extend and override *keywords*.
   Roughly equivalent to::

>       Placeholder = object()
> 
>       def partial(func, /, *args, **keywords):
>           # Trim trailing placeholders and count how many remaining
>           nargs = len(args)
>           while nargs and args[nargs-1] is Placeholder:
>               nargs -= 1
>           args = args[:nargs]
>           placeholder_count = args.count(Placeholder)
>           def newfunc(*fargs, **fkeywords):
>               if len(fargs) < placeholder_count:
>                   raise TypeError("missing positional arguments in 'partial' call;"
>                                   f" expected at least {placeholder_count}, "
>                                   f"got {len(fargs)}")
>               newargs = list(args)
>               j = 0
>               for i in range(len(args)):
>                   if args[i] is Placeholder:
>                       newargs[i] = fargs[j]
>                       j += 1
>               newargs.extend(fargs[j:])
>               newkeywords = {**keywords, **fkeywords}

Reviewer 3: ```suggestion
      Placeholder = object()

      def partial(func, /, *args, **keywords):
          # Trim trailing placeholders and count how many remaining
          nargs = len(args)
          while nargs and args[nargs-1] is Placeholder:
              nargs -= 1
          args = args[:nargs]
          placeholder_count = args.count(Placeholder)
          def newfunc(*fargs, **fkeywords):
              if len(fargs) < placeholder_count:
                  raise TypeError("missing positional arguments in 'partial' call;"
                                  f" expected at least {placeholder_count}, "
                                  f"got {len(fargs)}")
              newargs = list(args)
              j = 0
              for i, arg in enumerate(args):
                  if arg is Placeholder:
                      newargs[i] = fargs[j]
                      j += 1
              newargs.extend(fargs[j:])
              newkeywords = {**keywords, **fkeywords}
```

   Roughly equivalent to::

      def partial(func, /, *args, **keywords):
          def newfunc(*more_args, **more_keywords):

>               return func(*args, *more_args, **keywords, **more_keywords)

Reviewer: This is wrong.

The difference is that `{**keywords, **fkeywords}` allows duplicates, but in function call this is an error.

   The :func:`partial` is used for partial function application which "freezes"
   some portion of a function's arguments and/or keywords resulting in a new object

>    with a simplified signature. For example, :func:`partial` can be used to create

Reviewer 4: Nit: no need for this to change here :)

      >>> basetwo('10010')
      18

>    If ``Placeholder`` sentinels are present in *args*, they will be filled first

Reviewer 2: The above rough equivalent code needs to be updated.

>    If ``Placeholder`` sentinels are present in *args*, they will be filled first

Reviewer 5: ```suggestion
   If :class:`Placeholder` sentinels are present in *args*, they will be filled first
```

Author: As double back-ticked wrap is used for None in docs, I think this is not applicable anymore.

   when :func:`partial` is called. This allows custom selection of positional arguments
   to be pre-filled when constructing :ref:`partial object<partial-objects>`.

>    If ``Placeholder`` sentinels are used, all of them must be filled at call time.:

Reviewer 5: ```suggestion
   If :class:`!Placeholder` sentinels are used, all of them must be filled at call time.:
```

      >>> from functools import partial, Placeholder
      >>> say_to_world = partial(print, Placeholder, 'world!')
      >>> say_to_world('Hello')
      Hello world!

   .. versionchanged:: 3.14

>       Support for ``Placeholder`` in *args*

Reviewer 5: ```suggestion
      Support for :class:`!Placeholder` in *args*.
```

      >>> basetwo('10010')
      18

   If :data:`Placeholder` sentinels are present in *args*, they will be filled first
   when :func:`partial` is called. This allows custom selection of positional arguments
   to be pre-filled when constructing a :ref:`partial object <partial-objects>`.

   If :data:`!Placeholder` sentinels are used, all of them must be filled at call time:

>       >>> from functools import partial, Placeholder

Reviewer 3: Maybe use the `Placeholder as _` (since your other example also uses `_`).

Author: I think it is a good practice not to redefine `Placeholder` to a new name if not necessary. E.g. for 1 argument, one should just use full name and only re-define it when there are a lot of them for convenience and readability.

      >>> say_to_world = partial(print, Placeholder, Placeholder, "world!")
      >>> say_to_world('Hello', 'dear')
      Hello dear world!

   Calling ``say_to_world('Hello')`` would raise a :exc:`TypeError`, because
   only one positional argument is provided, while there are two placeholders
   in :ref:`partial object <partial-objects>`.

>    When successively using :func:`partial` existing :data:`!Placeholder`

Reviewer 3: ```suggestion
   When using :func:`partial` successively, existing :data:`!Placeholder`
```

   sentinels are filled first. A place for positional argument is retained

>    when :data:`!Placeholder` sentinel is replaced with a new one:

Reviewer 3: I don't understand the sentence "A place for positional argument is retained
   when <code>:data:\`!Placeholder\`</code> sentinel is replaced with a new one". Do you mean that you can still use placeholders if you don't want (yet) to fill them?

Author: It simply means that if there was a sentinel in input partial object, one can propagate it by overriding it with the new one.

I made some changes in that whole area. Hopefully it will be better.

      >>> from functools import partial, Placeholder as _
      >>> count = partial(print, _, _, _, 4)
      >>> count = partial(count, _, _, 3)
      >>> count = partial(count, _, 2)

>       >>> count = partial(count, _, 5)   # 5 is appended after 4

Reviewer 3: Just to confirm:

```py
# partial(print, _, _, _, 4)
def f1(x1, x2, x3, *args):
	print(x1, x2, x3, 4, *args)		# x1, x2, x3, 4, args[0], ...

# partial(f1, _, _, 3)
def f2(x1, x2, *args):
	f1(x1, x2, 3, *args)			# x1, x2, 3, 4, args[0], ...

# partial(f2, _, 2)
def f3(x1, *args):
	f2(x1, 2, *args)				# x1, 2, 3, 4, args[0], ...

# partial(f3, _, 5)
def f4(x1, *args):
	f3(x1, 5, *args)				# x1, 2, 3, 4, 5, args[0], ...
```

By the way, could you perhaps use `f` instead of `count`? (I don't feel it's *counting* something, it's more *printing* something).

Author: Yes, exactly. (to be precise, **kwds are also supported)

I think it is a matter of perspective. To me it is an object which after a lot of struggle eventually manages to count to 5. Not in a sense of object counting, but rather counting for no reason.

But ok, will change it to something else, maybe it can be confusing.

Reviewer 3: I can understand the reason, but `count` was a bit confusing for me (maybe I'm the only one that was confused though...)

Author: No, I can totally see your POV and I would guess more people would see the way you do.

      >>> count(1)
      1 2 3 4 5

   .. versionchanged:: 3.14

>       Support for :data:`Placeholder` in *args*

Reviewer 3: ```suggestion
      Added support for :data:`Placeholder` in positional arguments.
```

   If :data:`!Placeholder` sentinels are used, all of them must be filled at call time:

      >>> from functools import partial, Placeholder
      >>> say_to_world = partial(print, Placeholder, Placeholder, "world!")
      >>> say_to_world('Hello', 'dear')
      Hello dear world!

   Calling ``say_to_world('Hello')`` would raise a :exc:`TypeError`, because
   only one positional argument is provided, while there are two placeholders
   in :ref:`partial object <partial-objects>`.

   When successively using :func:`partial` existing :data:`!Placeholder`
   sentinels are filled first. A place for positional argument is retained
   when :data:`!Placeholder` sentinel is replaced with a new one:

      >>> from functools import partial, Placeholder as _
      >>> count = partial(print, _, _, _, 4)
      >>> count = partial(count, _, _, 3)
      >>> count = partial(count, _, 2)
      >>> count = partial(count, _, 5)   # 5 is appended after 4
      >>> count(1)
      1 2 3 4 5

   .. versionchanged:: 3.14
      Support for :data:`Placeholder` in *args*

.. data:: Placeholder

   A singleton object used as a sentinel to represent a

>    "gap" in positional arguments when calling :func:`partial`.

Reviewer: and partialmethod.

   in :ref:`partial object <partial-objects>`.

   Successive :func:`partial` applications fill :data:`!Placeholder` sentinels
   of the input :func:`partial` objects with new positional arguments.
   A place for positional argument can be retained by inserting new
   :data:`!Placeholder` sentinel to the place held by previous :data:`!Placeholder`:

      >>> from functools import partial, Placeholder as _
      >>> remove = partial(str.replace, _, _, '')
      >>> message = 'Hello, dear dear world!'
      >>> remove(message, ' dear')
      'Hello, world!'
      >>> remove_dear = partial(remove, _, ' dear')
      >>> remove_dear(message)
      'Hello, world!'
      >>> remove_first_dear = partial(remove_dear, _, 1)
      >>> remove_first_dear(message)
      'Hello, dear world!'

   Note, :data:`!Placeholder` has no special treatment when used for keyword
   argument of :data:`!Placeholder`.

   .. versionchanged:: 3.14
      Added support for :data:`Placeholder` in positional arguments.

.. data:: Placeholder

   A singleton object used as a sentinel to reserve a place
   for positional arguments when calling :func:`partial`

>    and :func:`partialmethod`.

Reviewer 4: This will need a `.. versionadded:: 3.14` tag.

Nit: it should also move to a better sort position, either logically before `partial` or alphabetically after `partialmethod` rather than sandwiched between the two.

   only one positional argument is provided, while there are two placeholders
   in :ref:`partial object <partial-objects>`.

   Successive :func:`partial` applications fill :data:`!Placeholder` sentinels
   of the input :func:`partial` objects with new positional arguments.
   A place for positional argument can be retained by inserting new
   :data:`!Placeholder` sentinel to the place held by previous :data:`!Placeholder`:

      >>> from functools import partial, Placeholder as _
      >>> show5 = partial(print, _, _, _, 4)
      >>> show5 = partial(show5, _, _, 3)
      >>> show5 = partial(show5, _, 2)

>       >>> show5 = partial(show5, _, 5)   # 5 is appended after 4

Reviewer: I am not sure that it is worth to keep such weird example, but I would use different names:
```
      >>> show4 = partial(print, _, _, _, 4)
      >>> show3 = partial(show4, _, _, 3)
      >>> show2 = partial(show3, _, 2)
      >>> show5 = partial(show2, _, 5)   # 5 is appended after 4
```
Well, it still does not make sense.

If you want more realistic examples, I can propose something like:
```
>>> remove = partial(str.replace, _, _, '')
>>> remove('Hello, world', 'l')
'Heo, word'
>>> remove_l = partial(remove, _, 'l')
>>> remove_l('Hello, world')
'Heo, word'
>>> remove_first_l = partial(remove_l, _, 1)
>>> remove_first_l('Hello, world')
'Helo, world'
```

Author: Thank you, much better.

      >>> basetwo('10010')
      18

   If ``Placeholder`` sentinels are present in *args*, they will be filled first
   when :func:`partial` is called. This allows custom selection of positional arguments
   to be pre-filled when constructing :ref:`partial object<partial-objects>`.

>    If ``Placeholder`` sentinels are used, all of them must be filled at call time.:

Reviewer 3: ```suggestion
   If ``Placeholder`` sentinels are used, all of them must be filled at call time::
```

>       >>> from functools import partial, Placeholder

Reviewer 3: I personally prefer having `PLACEHOLDER` since it's more a module constant rather than a global constant like `None`. In addition, I feel like it would stress the fact that it is a constant and not a class.

Author: It kinda follows the pattern of `None`, even if it is not in global namespace. However, I do uppercase my sentinels personally, so I wouldn't mind too much.

If there are no strong opinions, I would leave it as it is. After all, none of python's constants of this type are uppercased.

But I am open here if there is something that indicates uppercased would be better.

Reviewer 3: In light of your comment (https://github.com/python/cpython/pull/119827#issuecomment-2155989107), let's keep the PascalCase form (you can mark this discussion as resolved, at least from my side).

      >>> say_to_world = partial(print, Placeholder, 'world!')
      >>> say_to_world.placeholder_count
      1
      >>> say_to_world('Hello')
      Hello world!

   .. versionchanged:: 3.14

>       Support for ``Placeholder`` in *args*

Reviewer 3: You should document this additional constant that is publicly available, and then use <code>:data:\`~functools.Placeholder\`</code> to refer to it.

      >>> basetwo('10010')
      18

   If :data:`Placeholder` sentinels are present in *args*, they will be filled first
   when :func:`partial` is called. This allows custom selection of positional arguments
   to be pre-filled when constructing :ref:`partial object<partial-objects>`.

>    If :data:`Placeholder` sentinels are used, all of them must be filled at call time:

Reviewer 3: ```suggestion
   to be pre-filled when constructing a :ref:`partial object <partial-objects>`. 
   
   If :data:`Placeholder` sentinels are used, all of them must be filled at call time:
```

I would like you to also add a whitespace before the `<` in the paragraph just after `.. function:: partial` (I cannot point a non-modified line on a review).

>    If :data:`Placeholder` sentinels are used, all of them must be filled at call time:

Reviewer 5: We try to keep it to one unique ref pr. paragraph, in order to reduce the visual clutter.

```suggestion
   If :data:`!Placeholder` sentinels are used, all of them must be filled at call time:
```

      >>> from functools import partial, Placeholder
      >>> say_to_world = partial(print, Placeholder, Placeholder, "world!")
      >>> say_to_world.placeholder_count
      2
      >>> say_to_world('Hello', 'dear')
      Hello dear world!

>    Calling ``say_to_world('Hello')`` would result in :exc:`TypeError`.

Reviewer 3: ```suggestion
   Calling ``say_to_world('Hello')`` would raise :exc:`TypeError`.
```

   .. versionchanged:: 3.14
      Support for :data:`Placeholder` in *args*

.. data:: Placeholder

   This is a singleton object that is used as a sentinel for representing a
   gap in positional arguments when calling :func:`partial`.

>    It has similar features as :data:`None`:

Reviewer 3: ```suggestion
   A singleton object used as a sentinel to represent a "gap" 
   in positional arguments when calling :func:`partial`. This
   object has features simlar to ``None``:
```

I don't think we need an explicit link to `None` here.

Reviewer 5: - Please do not dismiss comments without providing a reason.
- Please read the devguide regarding how to [properly mark up None](https://devguide.python.org/documentation/markup/).
- I think the wording suggested by @picnixz should be considered.

Author: I am not dismissing. If I dismissed, it means I am incorporating it fully. I just do not press "Commit suggestion" as I don't like doing it this way.

Reviewer 5: In that case, it is nice to add a "Addressed in commit <Git hash>" comment.

   :data:`!Placeholder` sentinel to the place held by previous :data:`!Placeholder`:

      >>> from functools import partial, Placeholder as _
      >>> remove = partial(str.replace, _, _, '')
      >>> remove('Hello, world', 'l')
      'Heo, word'

>       >>> remove_l = partial(remove, _, 'l')

Reviewer: Actually, this may be not so good example, because `l` can be confused with `1` or `I`. Could you try to find better string for example?

`str.replace` is perhaps the best example, because we need a function that accepts at least 3 (preferably 4) positional arguments.

Author: Will do. Nitpicks are good, just maybe not very productive for that specific area that is going to change soon. Thanks!
